lightning-invoice/src/payment.rs

```diff
@@ -396,8 +396,8 @@ where
 		}.map_err(|e| PaymentError::Sending(e))
 	}
 
-	fn retry_payment(
-		&self, payment_id: PaymentId, payment_hash: PaymentHash, params: &RouteParameters
+	fn retry_payment(&self, payment_id: PaymentId, payment_hash: PaymentHash,
+		params: &RouteParameters, avoid_scid: Option<u64>
```

Comment:
I wonder if this would be cleaner if part of the `RouteParameters`.

Reply:
And avoided directly in `find_route`, that is.

Reply:
You mean passing it through to the router itself and asking it to completely avoid an SCID? That feels like its better done via the `Score` implementer, which I guess is ultimately the problem here - that the `Scorer` in use int he sample (ie our default one) doesn't strictly refuse to pay over a channel that just failed. That said, I do feel like the `InvoicePayer` should be robust against a braindead scorer, whether its our own or a user-provided one, so it feels nice to have it here too?

Reply:
Yeah, but also don't put the onus on the event handler to set it and pass it to `find_route`. Simply have the `ChannelManager` set it when creating the `PaymentPathFailed` event. Then it is completely transparent to anyone handling the event.

Reply:
Right, I do feel like "avoid this channel" is really more of a `Score` thing than a router thing - we have a whole interface for it, it seems annoying to duplicate that interface here. Its not a lot of code change, but still awkward.

Reply:
Hmm... but this use case is (a) ephemeral as it only applies to a specific payment -- lower payment amounts may be successful for another payment or even the failed path if further split on retry -- and (b) being handled by the caller not the scorer in this PR.

Reply:
Right in this instance its strangely dual-caller-scorer handling it - the scorer de-prioritizes and the caller handles the "oh, this went wrong, we cant do this, scorer or router are busted" case. I guess two more practical questions on behavior that may inform this more:

a) do we want to track this information across payment attempts - if there's two available last-hop hints do we want to just go back and forth between them until we run out of attempts,
b) do we care about avoiding the path in the router or are we okay with failing if we find the same path again (ie if the scorer is broken or doesn't learn, are we okay just failing the payment vs making sure the router picks another path)?

Both imply that the data should be in the `RouteParameters`, I think, if we care about either (I'm not sure we do), but (a) implies it should be in the `Payee` (to be renamed) not `RouteParameters`, even.

Reply:
Discussed this more offline, sounds like we want to/should go with moving the logic as described here, will do.